nntrainer/models/dynamic_training_optimization.h

```diff
@@ -0,0 +1,183 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Parichay Kapoor <pk.kapoor@samsung.com>
+ *
+ * @file   dynamic_training_optimization.h
+ * @date   4 January 2021
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Parichay Kapoor <pk.kapoor@samsung.com>
+ * @bug    No known bugs except for NYI items
+ * @brief  This is Dynamic Training Optimization for Neural Network
+ *
+ */
+
+#ifndef __DYNAMIC_TRAINING_OPT_H__
+#define __DYNAMIC_TRAINING_OPT_H__
+#ifdef __cplusplus
+
+#include <random>
+#include <vector>
+
+#include <layer_internal.h>
+#include <tensor.h>
+
+namespace nntrainer {
+
+/**
+ * @class   DynamicTraining Optimization
+ * @brief   Dynamic Training Optimization
+ */
+class DynamicTrainingOptimization {
+public:
+  /**
+   * @brief     Constructor of DynamicFineTuning Optimization
+   */
+  DynamicTrainingOptimization(int threshold_ = 1, int skip_n_iter = 1);
+
+  /**
+   * @brief     Set threshold for optimization
+   */
+  void setThreshold(float threshold_) { threshold = threshold_; };
+
+  /**
+   * @brief     Set the reduce operation for dynamic optimization
+   */
+  void setOp(std::string op) {
+    if (op == dft_opt_max)
+      reduce_op = reduceByMax;
+    else if (op == dft_opt_norm)
+      reduce_op = reduceByNorm;
+    else
+      enabled = false;
+  };
+
+  /**
+   * @brief     Enable the optimization
+   */
+  void enable() { enabled = true; }
+
+  /**
+   * @brief     Disable the optimization
+   */
+  void disable() { enabled = false; }
+
+  /**
+   * @brief     Set the mode for optimization
+   */
+  void setMode(std::string mode_) {
+    calc_ratio_mode = mode_;
+    if (mode_ == dft_opt_mode_derivative)
+      calc_ratio_op = ratioUsingDerivative;
+    else if (mode_ == dft_opt_mode_gradient)
+      calc_ratio_op = ratioUsingGradient;
+    else
+      enabled = false;
+  }
+
+  /**
+   * @brief     Check if the derivative mode is used for optimization
+   */
+  bool isDerivativeMode() {
+    if (enabled && calc_ratio_mode == dft_opt_mode_derivative)
+      return true;
+    return false;
+  }
+
+  /**
+   * @brief     Check if the gradient mode is used for optimization
+   */
+  bool isGradientMode() {
+    if (enabled && calc_ratio_mode == dft_opt_mode_gradient)
+      return true;
+    return false;
+  }
+
+  /**
+   * @brief     Set initial iterations to skip from optimization
```

Comment:
One simple question.
Does this variable means that at first n_iteration it want to skip the gradient or apply the gradient?

Reply:
for `first_n_iteration`, dynamic training optimization will be skipped, and gradients will always be applied.
Updated description to be clearer.